### DISCUSSION TITLE
How to implement FromRequest for    pub struct Validated<T>(pub Json<T>)

### DISCUSSION THREAD

**Discussion by zishon** [tokio-rs/axum#2793] [Q&A] [answered]
### Summary

I use validator to validate my every request struct, such as `CreateUserRequest`, `UpdateUserRequest`
so I make the hander like this: 

```rust
pub async fn create(Json(payload): Json<CreateUserRequest>) -> Json<CreateUserResponse> {
    //payload.validate() and some logic on validate result
}

pub async fn update(Json(payload): Json<UpdateUserRequest>) -> Json<UpdateUserResponse> {
    //payload.validate() and some logic on validate result
}
```

I don't want to do so everytime after adding new handler, so I wonder whether there is a way to do this all in one extractor with Generics, such as : pub struct Validated<T>(pub Json<T>)?

and the handler looks like 
```rust
pub async fn update(Validated(Json(payload)): Validated<Json<UpdateUserRequest>>) -> Json<UpdateUserResponse> {}
```

Thanks everyone, appreciate you all.

### axum version

0.7.5

**Comment by mladedav** [comment]
Did you try
```rust
impl<T> FromRequest for Validated<T> {
    // ...
}
```

and implementing the required method? Was there some issue with it?

I'm not sure though if an impl like that would actually help you since you can't have different validations for different payloads that way and thus you can check just common things (e.g. common fields are present). So you might actually want to do something like
```rust
impl FromRequest for Validated<CreateUserRequest> {
    // ...
}

impl FromRequest for Validated<UpdateUserRequest> {
    // ...
}
```

**Reply by zishon** [reply]
Yes, I had try as your way.
But it will be so much boilerplate code
```rust
impl FromRequest for Validated<CreateUserRequest> {
    // ...
}

impl FromRequest for Validated<UpdateUserRequest> {
    // ...
}
```

In my opinion, it would be nice to support something like this,
```rust
pub fn async create(Validated(Json(payload)): Validated<Json<CreateRequest>>) {}

//the only thing need to be satisfied is: SomeType derive Deserialize
pub fn async xxx(Validated(Json(payload)): Validated<Json<SomeType>>) {} 
```

**Reply by mladedav** [reply]
That looks like it would have about the same amount of boilerplate to me. You would have to create a method per struct that you wanted to validate. Which you're going to have to do anyway if the validation differs between structs, right?

Maybe I don't understand entirely what you're trying to achieve, can you try providing a more complete example of how the usage should look like?

**Reply by zishon** [reply, marked as answer]
maybe something like this
```rust
impl IntoResponse for RequestValidateErrors {
    //some implements
}
#[derive(Debug, Serialize, Deserialize)]
pub struct Validated<T>(pub T);

impl<S, B, T> FromRequest<B> for Validated<T>
    where
        B: Send + 'static,
        T: DeserializeOwned + Debug + Validate + FromRequest<S, B>,
        S: Send + Sync,
{
    type Rejection = RequestValidateErrors;

    async fn from_request(req: Request<Body>, state: &S) -> Result<Self, Self::Rejection> {
        let bytes = axum::body::to_bytes(req.into_body(), 100000).await.unwrap();

        let value = serde_json::from_slice::<T>(&bytes).unwrap();
        
        validate_res = value.validate();
        //process validate result, and if errors happened return RequestValidateErrors

        Ok(Validated(value))
    }
}
```
After that, all the error about Request extrators are deal in Validated<T>
Now the handler will focus on the business logic, of course, the extractor struct must be define in validator way
```rust
#[derive(Deserialize, Serialize, Validate, Debug)]
pub struct RegisterRequest {
    #[validate(length(min=6, max=16, code="1001", message="the length of username must be between 6 and 16"))]
    pub username: String,
}

pub async fn register(Validated(payload): Validated<RegisterRequest>) -> Result<GeneralResponse<RegisterResponse>, AppError> {
    //here, don't need to care about request validation
}
```

Really appreciate, thank you

**Reply by jplatte** [reply]
@zishon I wanted to suggest something like what you wrote in your last comment, probably call it `ValidatedJson`. So you got it to work and this question can be marked as answered?

**Reply by jplatte** [reply]
For some more prior art, you can also look into [axum-garde](https://docs.rs/axum_garde/latest/axum_garde/), which integrates the `garde` crate, a fork of `validator`, with axum.

**Reply by zishon** [reply]
> @zishon I wanted to suggest something like what you wrote in your last comment, probably call it `ValidatedJson`. So you got it to work and this question can be marked as answered?

Yes, exactly as you said.

**Reply by jplatte** [reply]
Okay, I marked your comment as answering the question. I think you could have also done that yourself, by the way.

**Reply by zishon** [reply]
> Okay, I marked your comment as answering the question. I think you could have also done that yourself, by the way.

It is what I want, but I haven't figure it out yet.
